Improve bookshelf.d.ts and tests

tkrotoff opened the pull request:
- More tests
- Switch to ES6 imports
- Add attributes and length
- Fix destroy()
- Fix detach()
- Fix map()
- Fix dropTable()
- Add NotFoundError, NoRowsUpdatedError and NoRowsDeletedError on Model

arcticwaters commented on bookshelf/bookshelf.d.ts:
```diff
@@ -177,8 +191,12 @@ declare module 'bookshelf' {
-			map(predicate? : Lodash.ListIterator<T, boolean>|Lodash.DictionaryIterator<T, boolean>|string, thisArg? : any) : T[];
+
+			// See https://github.com/Microsoft/TypeScript/blob/v1.8.10/lib/lib.core.es7.d.ts#L1122
+			map<U>(predicate? : Lodash.ListIterator<T, U>|string, thisArg? : any) : U[];
```
Why drop `DictionaryIterator`?

tkrotoff replied:
Very good question.

See the test here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/9297/files#diff-8a3091671ada9e45452b597b82364fddR214, line:

```JavaScript
const booksOutput = books.map(book => new Author2Output(book.toJSON()));
```

`book` should be of type `Book` (and `booksOutput` of type `AuthorOutput[]`)
- works with `map<U>(predicate? : Lodash.ListIterator<T, U>|string, thisArg? : any) : U[];`
- works with `map<U>(predicate? : Lodash.DictionaryIterator<T, U>|string, thisArg? : any) : U[];`
- does not work (`book` is of type `any`) with `map<U>(predicate? : Lodash.ListIterator<T, U>|Lodash.DictionaryIterator<T, boolean>|string, thisArg? : any) : U[];`

TypeScript is lost with `Lodash.ListIterator<T, U>|Lodash.DictionaryIterator<T, U>`, I don't know why.

<br>

This:
```JavaScript
map<U>(predicate? : Lodash.ListIterator<T, U>|string, thisArg? : any) : U[];
map<U>(predicate? : Lodash.DictionaryIterator<T, U>|string, thisArg? : any) : U[];
```
works => the best solution?

I don't know the differences between `ListIterator` and `DictionaryIterator`, not a lodash expert :)

arcticwaters replied:
The difference is `iter[2]` and `iter['two']`. See https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/lodash/lodash-3.10.d.ts#L15807. My only concern is maintaining lodash APIs even though I may not use them. I wish there as a better way to do that.

You are right in your assumption though. I don't know why typescript doesn't work either. My understanding of typescript says that (3) and the last are identical, but if only the last works (or is easier to write), then the last may be the best solution. If there is a reason why lodash splits the methods up instead of combining them like I have, I'm unsure of the why. Perhaps we should split them up identical to the lodash methods, i.e.
```
map<U>(predicate? : Lodash.ListIterator<T, U>, thisArg? : any) : U[];
map<U>(predicate? : Lodash.Dictionary<T, U>, thisArg? : any) : U[];
map<U>(predicate? : string) : U[];
```
(I just realized that a string predicate doesn't have a this parameter). This might be why sometimes I see `map<Book>(...)` instead of `map(...)`.

arcticwaters commented on bookshelf/bookshelf.d.ts:
```diff
@@ -177,8 +191,12 @@ declare module 'bookshelf' {
+			// See https://github.com/Microsoft/TypeScript/blob/v1.8.10/lib/lib.core.es7.d.ts#L1122
+			map<U>(predicate? : Lodash.ListIterator<T, U>|string, thisArg? : any) : U[];
+			//map(predicate? : Lodash.ListIterator<T, boolean>|Lodash.DictionaryIterator<T, boolean>|string, thisArg? : any) : T[];
 			map<R extends {}>(predicate? : R) : T[];
```
Can we fix this map method too? It should probably return `U[]` as the one above it.

tkrotoff replied:
Are you talking about `map<R extends {}>(predicate? : R) : T[];`?
I don't know what this line does: I don't know how to write a test/example for it, that's why I didn't touch it.

arcticwaters replied:
Yup. I suspect this is to handle the case of objects which don't have an explicit class/interface definition. Again, this method is just there to satisfy the lodash API.

arcticwaters commented on bookshelf/bookshelf.d.ts:
```diff
@@ -177,8 +191,12 @@ declare module 'bookshelf' {
 			lastIndexOf(value : any, fromIndex? : number) : number;
-			map(predicate? : Lodash.ListIterator<T, boolean>|Lodash.DictionaryIterator<T, boolean>|string, thisArg? : any) : T[];
+
+			// See https://github.com/Microsoft/TypeScript/blob/v1.8.10/lib/lib.core.es7.d.ts#L1122
```
You should probably reference the lodash docs if you want to reference something: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/lodash/lodash-3.10.d.ts#L7119 or https://github.com/lodash/lodash/blob/3.10.1/lodash.src.js#L6864

tkrotoff replied:
See the `map<U>(...): U[] `? I took the **U** solution/idea from https://github.com/Microsoft/TypeScript/blob/v1.8.10/lib/lib.core.es7.d.ts#L1122:

```JavaScript
map<U>(callbackfn: (value: T, index: number, array: T[]) => U, thisArg?: any): U[];
```

`ListIterator` from lodash being:
```JavaScript
interface ListIterator<T, TResult> {
  (value: T, index: number, collection: List<T>): TResult;
}
```
So the 'same' as `(value: T, index: number, array: T[]) => U` from lib.core.es7.d.ts

That's why I put a link to lib.core.es7.d.ts (and it's a good idea to link to lodash-3.10.d.ts#L7119 too).

arcticwaters replied:
You are right referencing a new `U` parameter. I don't know why I referenced `boolean` in the initial implementation. My thought was that it might be confusing to reference this es7 method since the lodash method overrides it (even though the API is almost identical).

arcticwaters commented:
Can we add `EmptyError` to `Collection`? It's the only one that's missing now.

tkrotoff commented:
> Can we add `EmptyError` to `Collection`?

I didn't because I don't have a test for it.

tkqubo commented:
For `knex` it LGTM, thanks 👍

Blackman21 commented:
Pls allow this request. I would like to use the changes :)

tkrotoff commented:
@arcticwaters 
I've added **all** examples from Bookshelf documentation. It allowed me to fix more things.

Regarding your remarks, I've fixed `map()` again and added `EmptyError`.

Thx for your help, the PR should be good now.

arcticwaters commented:
Cool. Looks good. :+1:

tkrotoff commented:
@arcticwaters 
I have a question about `through()`:
- [`Model.through<R ...>(...) : R | Collection<R>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/1ec3d510e0c4d7c50e708ca7cb0765c7e06fa2ed/bookshelf/bookshelf.d.ts#L109)

- [`Collection.through<R ...>(...) : R | Collection<R>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/1ec3d510e0c4d7c50e708ca7cb0765c7e06fa2ed/bookshelf/bookshelf.d.ts#L222)

I wonder if it's not instead:
- `Model.through<R ...>(...) : R`
- `Collection.through<R ...>(...) : Collection<R>`

This way it works beautifully on my app code + on Bookshelf examples:

```JavaScript
	class Doctor extends bookshelf.Model<Doctor> {
		patients(): Bookshelf.Collection<Patient> {
			return this.belongsToMany(Patient).through<Patient>(Appointment);
		}
	}

	class Patient extends bookshelf.Model<Patient> {
		doctors(): Bookshelf.Collection<Doctor> {
			return this.belongsToMany(Doctor).through<Doctor>(Appointment);
		}
	}

	class Book extends bookshelf.Model<Book> {
		// Find all paragraphs associated with this book, by
		// passing through the "Chapter" model.
		paragraphs(): Paragraphs {
			return this.hasMany(Paragraph).through<Paragraph>(Chapter);
		}
	}

	class Paragraph extends bookshelf.Model<Paragraph> {
		// A reverse relation, where we can get the book from the chapter.
		book(): Book {
			return this.belongsTo(Book).through<Book>(Chapter);
		}
	}
```

Also it fits well the unit tests from Bookshelf, examples:
- https://github.com/tgriesser/bookshelf/blob/0.9.4/test/integration/relation.js#L21
```JavaScript
      // belongsToMany() returns a Collection
      // patient*s*() returns a Collection<Patient> not a just a single Patient
      patients: function() {
        return this.belongsToMany(Patient).through(Appointment);
      }
```

- https://github.com/tgriesser/bookshelf/blob/0.9.4/test/integration/relation.js#L42
```JavaScript
      doctors: function() {
        return this.belongsToMany(Doctor).through(Appointment);
      }
```

- https://github.com/tgriesser/bookshelf/blob/0.9.4/test/integration/relation.js#L62
```JavaScript
      // hasOne() returns a Model
      accountHistory: function() {
        return this.hasOne(AccountHistory).through(Account);
      }
```

- https://github.com/tgriesser/bookshelf/blob/0.9.4/test/integration/relation.js#L73
```JavaScript
      // belongsTo() returns a Model
      supplier: function() {
        return this.belongsTo(Supplier).through(Account);
      }
```

- https://github.com/tgriesser/bookshelf/blob/0.9.4/test/integration/helpers/objects.js#L65
```JavaScript
    info: function() {
      return this.hasOne(Info).through(SiteMeta, 'meta_id');
    }
```

...

I could not find a counterexample.

arcticwaters commented:
You may be right. I can't think of a counterexample either. The [bookshelf docs](http://bookshelfjs.org/#Model-instance-through) say `model.through(...)` returns a collection, although they aren't the most complete or accurate.

There might be some odd case that we haven't thought of, but I'm happy to make this change until we figure out what that case is and why its important.

tkrotoff commented:
Here we go...

arcticwaters commented:
:+1:
